cuqi/implicitprior/_restorator.py

```diff
@@ -0,0 +1,218 @@
+    @property
+    def _mutable_vars(self):
+        """ Returns the mutable variables of the distribution. """
+        # Currently mutable variables are not supported for user-defined
+        # distributions.
+        return []
+
+    #TODO this method copied from userdefinedistribution
+    def get_conditioning_variables(self):
+        """ Returns the conditioning variables of the distribution. """
+        # Currently conditioning variables are not supported for user-defined
+        # distributions.
+        return []
+############################################################################################   
+class MoreauYoshidaPrior(Distribution):
+    """    
+    This class defines (implicit) smoothed priors for which we can apply 
+    gradient-based algorithms. The smoothing is performed using
+    the Moreau-Yoshida envelope of the target prior potential.
+    
+    In the following we give a detailed explanation of the
+    Moreau-Yoshida smoothing.
+    
+    We consider a density such that - \log\pi(x) = -g(x) with g convex, lsc,
+    proper but not differentiable. Consequently, we cannot apply any
+    algorithm requiring the gradient of g.
+    Idea:
+    We consider the Moreau envelope of g defined as
+    
+    g_{smoothing_strength} (x) = inf_z 0.5*\| x-z \|_2^2/smoothing_strength + g(z).
+    
+    g_{smoothing_strength} has some nice properties
+        - g_{smoothing_strength}(x)-->g(x) as smoothing_strength-->0 for all x
+        - \nabla g_{smoothing_strength} is 1/smoothing_strength-Lipschitz
+        - \nabla g_{smoothing_strength}(x) = (x - prox_g^{smoothing_strength}(x))/smoothing_strength for all x with 
+        
+        prox_g^{smoothing_strength}(x) = argmin_z 0.5*\| x-z \|_2^2/smoothing_strength + g(z) .
+
+    Consequently, we can apply any gradient-based algorithm with
+    g_{smoothing_strength} in lieu of g. These algorithms do not require the
+    full knowledge of g_{smoothing_strength} but only its gradient. The gradient
+    of g_{smoothing_strength} is fully determined by prox_g^{smoothing_strength}
+    and smoothing_strength.
+    It is important as, although there exists an explicit formula for
+    g_{smoothing_strength}, it is rarely used in practice, as it would require
+    us to solve an optimization problem each time we want to 
+    estimate g_{smoothing_strength}. Furthermore, there exist cases where we dont't
+    the regularization g with which the mapping prox_g^{smoothing_strength} is
+    associated.
+
+    Remark (Proximal operators are denoisers):
+        We consider the denoising inverse problem x = u + n, with
+        n \sim \mathcal{N}(0, smoothing_strength I).
+        We assume a prior density \pi(u) \propto exp(- g(u)).
+        Then the MAP estimate is given by 
+            x_MAP = \argmin_z 0.5 \| x - z \|_2^2/smoothing_strength + g(z) = prox_g^smoothing_strength(x) ()
+        Then proximal operators are denoisers. 
```

Comment:
If think for the statement that "proximal operators are denoisers" to make sense, we need a definition of a denoiser.

Reply:
Thank you Jakob. That is a good point. We added a definition of denoisers in the documentation of MoreauYoshidaPrior.